### ISSUE TITLE
The cursor does not disable active mode when I go to a new page in Next JS

### ISSUE THREAD

**Issue by ivanjeremic** [stephenscaff/react-animated-cursor#6]
The cursor does not disable active mode when I go to a new page in Next JS, It only works if I randomly click anywhere in the body on the new page then it works again, basically when I click a link and on the new page the cursor stays active the active does not go away until I click somewhere in the body. 

Any solution to this? The library needs somehow to detect a page change a reset everything in the cursor until it goes over a link again.

**Comment by JordanSuarez** [comment]
Hello, 

I'm not using Next JS, but i have the same issue, when i change page the Cursor does not disable active mode.

I found a solution, it's not the better of course, but it's work for me, it's just for a small project (my portfolio).

So the solution consist to import the Cursor component inside the router component (react router dom),  so I can use the useHistory hook to detect if the current location has change.

![Screenshot 2021-02-26 at 12 03 22](https://user-images.githubusercontent.com/67593837/109292422-a0ffa480-782a-11eb-9dcd-80ec499f21b3.png)

Then, inside the Cursor component, I added a useEffect which will update the Cursor state (isActive) at each change of route.

![Screenshot 2021-02-26 at 12 03 58](https://user-images.githubusercontent.com/67593837/109292472-b5dc3800-782a-11eb-9902-8a157f1971a1.png)

i hope it will help

**Comment by stephenscaff** [comment]
Sorry, just caught this. @ivanjeremic I assume you're using Next's Link helper when updating routes?
And thanks for the insight @JordanSuarez. I'll see what we can come with built in, but some kind of way to update if isActive feels like a good move. Or, maybe just going inactive after clicks?

Will check it out today.

**Comment by ivanjeremic** [comment]
> Sorry, just caught this. @ivanjeremic I assume you're using Next's Link helper when updating routes?
> And thanks for the insight @JordanSuarez. I'll see what we can come with built in, but some kind of way to update if isActive feels like a good move. Or, maybe just going inactive after clicks?
> 
> Will check it out today.

Yes, there needs to be something that sets the cursor inactive after clicks, I tried the solution @JordanSuarez suggested long before opening this issue and it would work in a normal CRA but in NextJS the method has no effect the cursor still stays active after clicking a link or button and navigating to a new page, any recommendations to quick fix this until there is something built in!

**Comment by ivanjeremic** [comment]
Anything new on this?

**Comment by baktiaditya** [comment]
hope can give you some inspiration:
https://gist.github.com/baktiaditya/41d83e04972cca950b59bb7ac83e7930#file-cursor-tsx-L62-L82

**Comment by ivanjeremic** [comment]
> hope can give you some inspiration, I'm using this in my Next.js project (modified version of react-animated-cursor)
> https://gist.github.com/baktiaditya/41d83e04972cca950b59bb7ac83e7930#file-cursor-tsx-L60-L81
> 
> usage in Next.js `_app.tsx`:
> 
> ```tsx
> const router = useRouter();
> const cursorRef = React.useRef<CursorRef>(null);
> 
> React.useEffect(() => {
>   router.events.on('routeChangeComplete', progressDone);
>   router.events.on('routeChangeError', progressDone);
> 
>   return () => {
>     router.events.off('routeChangeComplete', progressDone);
>     router.events.off('routeChangeError', progressDone);
>   };
> }, []);
> 
> const progressDone = () => {
>   const { current: cursor } = cursorRef;
>   if (cursor) {
>     cursor.update();
>   }
> };
> 
> return (
>   <Cursor ref={cursorRef} />
> );
> ```

Thank you so much!

**Comment by ivanjeremic** [comment]
@baktiaditya Thank you, do I still need to import it dynamically? It works perfectly for the built-in `a` tags but I have some dynamically generated elements that do not trigger the cursor animation even when added to the `clickableTargets` any idea?

**Comment by ivanjeremic** [comment]
@baktiaditya Hope you have some advice for me why some links only work after I save some work in the editor and instantly try on the Site but as soon as I refresh the browser it stops working until I add some random comment in my code and safe. hmm

**Comment by stephenscaff** [comment]
Sorry. Been crushed lately but have some time to look at this after
tomorrow.

Can you make a reduced test to look at? I haven’t had that issue.

**Comment by baktiaditya** [comment]
> `clickableTargets`



> @baktiaditya Thank you, do I still need to import it dynamically? It works perfectly for the built-in `a` tags but I have some dynamically generated elements that do not trigger the cursor animation even when added to the `clickableTargets` any idea? I use also a lot of 'id's in my targets can this be an issue?
> 
> It works only when I do some random changes in the editor and save and go directly over some element then the animation works, but as soon as I reload it stops working.

You can refactor my `Cursor` code and try to make the `clickableTargets` as a props. And for the dynamic elements, you can use react context to pass the `cursorRef` from the `_app.tsx` file, then use `cursor.update` when your dynamic elements is shown.

**Comment by ygorluiz** [comment]
> hope can give you some inspiration, I'm using this in my Next.js project (modified version of react-animated-cursor) https://gist.github.com/baktiaditya/41d83e04972cca950b59bb7ac83e7930#file-cursor-tsx-L60-L81
> 
> usage in Next.js `_app.tsx`:
> 
> ```tsx
> const router = useRouter();
> const cursorRef = React.useRef<CursorRef>(null);
> 
> React.useEffect(() => {
>   router.events.on('routeChangeComplete', progressDone);
>   router.events.on('routeChangeError', progressDone);
> 
>   return () => {
>     router.events.off('routeChangeComplete', progressDone);
>     router.events.off('routeChangeError', progressDone);
>   };
> }, []);
> 
> const progressDone = () => {
>   const { current: cursor } = cursorRef;
>   if (cursor) {
>     cursor.update();
>   }
> };
> 
> return (
>   <Cursor ref={cursorRef} />
> );
> ```

Can you tell me how do i add the magnetic effect like this `https://tympanus.net/Tutorials/CustomCursors/index3.html` ? BTW thx

**Comment by milindgoel15** [comment]
Any updates? It seems to be causing it on app router for me but fine on pages router

**Comment by stephenscaff** [comment]
Yikes…. Been under water with work stuff and slacked on this.

But have a better understanding of RSC and next14 app dir now.

So, will take a closer look, probs tomorrow (have a very sick kiddo right
now).

Can ping you once I push something up.

Thanks.

**Comment by ivanjeremic** [comment]
I solved this with a nice hack and fork, can publish my version to npm if so many people have this issue.

Gesendet von Outlook für iOS<https://aka.ms/o0ukef>
________________________________
Von: Stephen Scaff ***@***.***>
Gesendet: Monday, January 15, 2024 6:32:32 PM
An: stephenscaff/react-animated-cursor ***@***.***>
Cc: Ivan Jeremic ***@***.***>; Mention ***@***.***>
Betreff: Re: [stephenscaff/react-animated-cursor] The cursor does not disable active mode when I go to a new page in Next JS (#6)


Yikes…. Been under water with work stuff and slacked on this.

But have a better understanding of RSC and next14 app dir now.

So, will take a closer look, probs tomorrow (have a very sick kiddo right
now).

Can ping you once I push something up.

Thanks.

**Comment by stephenscaff** [comment]
Submit a PR if you’d like and I’ll merge it in.

**Comment by ivanjeremic** [comment]
I will dig it out from the codebase where I used my custom cursor and try to fix it here once I find time.

**Comment by adamgobes** [comment]
@ivanjeremic mind sharing how you solved the issue? I'm also running into it.

TLDR is the cursor stays active and enlarged when I click on a NextJS route and am navigated to a new page
